src/Npgsql/Internal/NpgsqlConnector.cs

```diff
@@ -839,10 +841,10 @@ async Task RawOpen(SslMode sslMode, NpgsqlTimeout timeout, bool async, Cancellat
                         certificateValidationCallback = SslTrustServerValidation;
                         checkCertificateRevocation = false;
                     }
-                    else if ((Settings.RootCertificate ?? PostgresEnvironment.SslCertRoot ?? PostgresEnvironment.SslCertRootDefault) is
-                             { } certRootPath)
+                    else if ((caCert = DataSource.CACertificateCallback?.Invoke()) is not null || (certRootPath = Settings.RootCertificate ?? PostgresEnvironment.SslCertRoot ?? PostgresEnvironment.SslCertRootDefault) is
```

Comment:
Should Settings.RootCertificate take precedence over Datasource.CACertificateCallback, since the latter is programmatic and the latter can be changed via config without recompiling? I think in general we allow connection string parameter to override programmatic setting.

Reply:
> I've assumed that in case you set the callback you have some sort of an API to provide the cert via a config. And in that case you probably want to use that instead of environment variables.

We should probably decide on these rules once and for all... It feels like we've had this discussion many times in various forms :)

> One way we can go about this is to use both certificates if found (we can add as many root certificates as we want for validation).

I guess so - that's what we already do for cilent certificates (merge the callback results with the connection string/environment). But if we support multiple root certs, we should also add an API for that directly on NpgsqlDataSourceBuilder (like how we already have UseClientCertificates which accepts a collection), no?

Reply:
> I guess so - that's what we already do for cilent certificates (merge the callback results with the connection string/environment). But if we support multiple root certs, we should also add an API for that directly on NpgsqlDataSourceBuilder (like how we already have UseClientCertificates which accepts a collection), no?

We can do that, though it's hard for me to imagine having multiple root certificates (you're supposed to know what certificate is used as a root and it should be the only one). The main point was for us to not make the decision about the certificate ordering...

Reply:
Sure thing. If the need ever comes up, we can always add another overload accepting multiple certificates without it being a breaking change.